File: Python3/intersection_of_two_linked_lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def getIntersectionNode(self, headA: ListNode, headB: ListNode) -> ListNode:
        hashmap = {}
        
        while headA:
            hashmap[headA] = headA.val
            headA = headA.next
            
        while headB:
            if headB in hashmap: return headB
            headB = headB.next
            
        return None

File: C++/intersection_of_two_linked_lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *getIntersectionNode(ListNode *headA, ListNode *headB) {
        if(headA == NULL or headB == NULL) return NULL;
        
        ListNode* path_A = headA;
        ListNode* path_B = headB;
        
        while(path_A != path_B) {
            if(path_A == NULL) path_A = headB; 
            else path_A = path_A->next; 
            
            if(path_B == NULL) path_B = headA; 
            else path_B = path_B->next;
        }        
        return path_A;
    }
};
